Update retention documentation for 2.0

WireBaron opened the pull request:
This change modifies several of the documentation pages to reflect
the api changes introduced in 2.0. Most significantly, removing the
cascade_to_materializations option from drop chunks and the
associated policy invalidated a large section of documentation we
had explaining it.  This has been replaced with warnings of
potential interactions between continuous aggregates and drop_chunks
and instructions for how to avoid these issues.

Additionally, there was a section on using external job schedulers
within the data-retention page. Given that the retention policy
framework should now be a preferable approach for almost all users,
this section was removed.

PR instructions

1. Describe your PR right below:



---
2. Add label for the earliest DB version your changes apply to -->>

3. If you are changing page names or in-page anchors, you must update the
`page-index.js` file

4. Please add at least one content reviewer. If you don't add an
additional reviewer with knowledge about the area you are writing about,
one of the codeowners may add one

5. If this is in response to a posted GitHub issue, please add "Fixes <issue #>"
below so that it's referenced (i.e. "Fixes #122")

5. If you are making simple corrections, reviewers may add comments to your
PR without officially requesting changes. This is to avoid additional
request/review cycles for simple changes. Make sure to address these
comments (i.e. with changes) before your PR is ready to merge

6. Feel free to delete these instructions in your PR message after everything
else is filled out
---
Additional notes:

The default reviewers (CODEOWNERS) are added to each PR.  They will
primarily be reviewing on formatting, not content.  You only need ONE of them
to approve in order to merge your PR

mkindahl commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -323,24 +323,19 @@ aggregate, such as views that are built on top of the continuous aggregate view.
-The same argument must also be supplied to the [`add_retention_policy`
-function][api-add-retention] when creating a data retention policy for a
-hypertable with a continuous aggregate.  
+Note that if any still-refreshing (more recent than start_offset) part of the
```
```suggestion
Note that if any still-refreshing (more recent than `start_offset`) part of the
```

mkindahl replied:
On a more general note: I think it is useful to introduce the concept of "invalidations" because that makes the rest of the text easier to follow and understand.

WireBaron replied:
I do think that invalidations are probably too much of an implementation specific detail here.  Users should be able to think of continuous aggregates as just continually updating within a specified region without needing to think about our behind the scenes bookkeeping.

mkindahl commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -323,24 +323,19 @@ aggregate, such as views that are built on top of the continuous aggregate view.
-The same argument must also be supplied to the [`add_retention_policy`
-function][api-add-retention] when creating a data retention policy for a
-hypertable with a continuous aggregate.  
+Note that if any still-refreshing (more recent than start_offset) part of the
+continuous aggregate is dropped via a [retention policy][api-add-retention] or
+direct [`drop_chunks`][api-drop-chunks] call, the aggregate will be updated to
+reflect the loss of data. For this reason, if it desired to retain the continuous
```
I think you need to have something along the lines "If there are outstanding invalidations in the hypertable, ....".

mkindahl replied:
... also, what happens if the user does an explicit `DELETE`? I assume nothing specific happens, but it is useful to mention to indirectly suggest that `drop_chunks` function should be used.

mkindahl commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
-Once these units have been created, run
+Extra care must be taken when using rention policies or `drop_chunks` calls on
+hypertables which have [continuous aggregates][continuous_aggregates] defined on
+them. Any chunks which are dropped in the region still being refreshed by the
```
This is a defining clause.
```suggestion
them. Any chunks that are dropped in the region still being refreshed by the
```

mkindahl commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
+continuous aggregate will cause the chunk data to disappear from the aggregate as
+well. If the intent is to keep the aggregate while dropping the underlying data,
+the interval being dropped should not overlap with the offsets for the continuous
+aggregate.  More detail on this can be found [here][cagg_drop_data].
```
```suggestion
aggregate.  More detail on this can be found in the section [whatever the name of the section is][cagg_drop_data].
```

mkindahl commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
+the interval being dropped should not overlap with the offsets for the continuous
+aggregate.  More detail on this can be found [here][cagg_drop_data].
+
+One additional item to note is that continuous aggregates are also valid targets
```
```suggestion
Also note that continuous aggregates are valid targets
```

mkindahl commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
-To make systemd load the timer unit, enable it by default on bootup,
-and immediately start it.
 
+Note that for this to work properly, it is important that the `end_offset` for
```
```suggestion
For this to work properly, it is important that the `end_offset` for
```

mfreed replied:
Should we show this explicitly in the example, then?  I'm trying to follow this, but with the changes I see proposed, I don't see the definition for conditions_summary_daily anywhere:

![image](https://user-images.githubusercontent.com/205086/94634671-a7e7b280-029e-11eb-83bc-f9ca3398bbb6.png)

WireBaron replied:
I've updated the example, hopefully it's much easier to follow now.

svenklemm commented on using-timescaledb/data-retention.md:
````diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
 The scheduler framework also allows one to view scheduled jobs:
 ```sql
-SELECT * FROM timescaledb_information.drop_chunks_policies;
+SELECT * FROM timescaledb_information.policy_stats;
````
This will probably be renamed to job_stats but final decission is still pending

mfreed replied:
Update to `job_stats`

mfreed commented on using-timescaledb/data-retention.md:
````diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
-```
-
-Once these units have been created, run
+Extra care must be taken when using rention policies or `drop_chunks` calls on
````
```suggestion
Extra care must be taken when using retention policies or `drop_chunks` calls on
```

erimatnor commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,38 @@ SELECT remove_retention_policy('conditions');
-Once these units have been created, run
+Extra care must be taken when using rention policies or `drop_chunks` calls on
+hypertables which have [continuous aggregates][continuous_aggregates] defined on
+them. Any chunks which are dropped in the region still being refreshed by the
```
```suggestion
them. Similar to a refresh of a materialized view, a refresh on a continuous aggregate will update the aggregate to reflect changes in the underlying source data. This means that any chunks which are dropped in the region still being refreshed by the
```

WireBaron replied:
I really liked how you captured this, added

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -323,24 +323,19 @@ aggregate, such as views that are built on top of the continuous aggregate view.
+
+This is also a consideration when manually refreshing a continuous aggregate.
+Calling `refresh_continuous_aggregate` on a region containing dropped chunks will
+recalculate the aggregate without the dropped data.
```
I think it is helpful to illustrate the correct usage (or the problem) with an example.

WireBaron replied:
I've added a more fully fleshed example in the data-retention document and added a link to that doc from here.

WireBaron commented:
As part of this most recent change, I've added a more in-depth example to the data-retention.md page and changed the continuous aggregates doc to point back to that.

mkindahl commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,58 @@ SELECT remove_retention_policy('conditions');
-
-[Timer]
-OnCalendar=*-*-* 03:00:00
+This will create the `conditions_summary_daily` aggregate which will store the daily
+temperature per device from our `conditions` table. However, we have a problem here
+if we're using our 24 hour rentention policy from above, as our aggregate will capture
+changes to the data for up to seven days. As a result, we will update the aggregate
+when we drop the chunk from the table, and we'll ultimately end up with no data in our
+`conditions_summary_daily` table.
```
Hmmm... usually it is not recommended to show incorrect examples since misunderstanding are quite common. It is usually more common to show a correct example and then digress into the problematic case: along the lines "Note that ... if we did it this way instead, it would not work since...". Then you keep the "faulty" code located to the discussion about it.

mkindahl reviewed (approved):
One comment about a restructuring, but since I think we can take it separately, I'm approving the PR.

mfreed commented on using-timescaledb/data-retention.md:
```diff
@@ -147,72 +69,58 @@ SELECT remove_retention_policy('conditions');
-OnCalendar=*-*-* 03:00:00
+This will create the `conditions_summary_daily` aggregate which will store the daily
+temperature per device from our `conditions` table. However, we have a problem here
+if we're using our 24 hour rentention policy from above, as our aggregate will capture
```
```suggestion
if we're using our 24 hour retention policy from above, as our aggregate will capture
```

WireBaron replied:
Fixed

mfreed commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -402,6 +402,22 @@ and information about the state and progress of the materialization background w
+
+This is also a consideration when manually refreshing a continuous aggregate.
+Calling `refresh_continuous_aggregate` on a region containing dropped chunks will
+recalculate the aggregate without the dropped data.
```
```suggestion
recalculate the aggregate without the dropped data (which can lead to undesirable
results, such as replacing previous aggregate data with NULL values, given that the
raw data has subsequently been dropped).
```

WireBaron replied:
Added this as a new sentence (`This can lead to ...`).

mfreed commented on using-timescaledb/data-retention.md:
````diff
@@ -147,72 +69,58 @@ SELECT remove_retention_policy('conditions');
-Description=Drop chunks from the 'conditions' table
+As an example, let's add a continuous aggregate to our `conditions` hypertable:
+```sql
+CREATE MATERIALIZED VIEW conditions_summary_daily (day, device, temp)
+WITH (timescaledb.continuous) AS
+SELECT time_bucket('1 day', time), device, AVG(temperature)
+FROM conditions
+GROUP BY (1, 2);
````
```suggestion
CREATE MATERIALIZED VIEW conditions_summary_daily (day, device, temp)
WITH (timescaledb.continuous) AS
  SELECT time_bucket('1 day', time), device, avg(temperature)
  FROM conditions
  GROUP BY (1, 2);
```

WireBaron replied:
fixed

mfreed commented:
This looks ready for merge to me.
